crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -100,55 +244,341 @@ fn site_packages_dir_from_sys_prefix(
+                    }
+                    "home" => base_executable_home_path = Some(value),
+                    // `virtualenv` and `uv` call this key `version_info`,
+                    // but the stdlib venv module calls it `version`
+                    "version" | "version_info" => version_info_string = Some(value),
+                    _ => continue,
+                }
+            }
+        }
+
+        // The `home` key is read by the standard library's `site.py` module,
+        // so if it's missing from the `pyvenv.cfg` file
+        // (or the provided value is invalid),
+        // it's reasonable to consider the virtual environment irredeemably broken.
+        let Some(base_executable_home_path) = base_executable_home_path else {
+            return Err(SitePackagesDiscoveryError::PyvenvCfgParseError(
+                pyvenv_cfg_path,
+                PyvenvCfgParseErrorKind::NoHomeKey,
+            ));
+        };
+        let Some(base_executable_home_path) =
+            PythonHomePath::new(base_executable_home_path, system)
+        else {
+            return Err(SitePackagesDiscoveryError::PyvenvCfgParseError(
+                pyvenv_cfg_path,
+                PyvenvCfgParseErrorKind::HomeValueIsNotADirectory,
+            ));
+        };
+
+        // but the `version`/`version_info` key is not read by the standard library,
+        // and is provided under different keys depending on which virtual-environment creation tool
+        // created the `pyvenv.cfg` file. Lenient parsing is appropriate here:
+        // the file isn't really *invalid* if it doesn't have this key,
+        // or if the value doesn't parse according to our expectations.
+        let minor_version = version_info_string.and_then(|version_string| {
+            let mut version_info_parts = version_string.split('.');
+            if version_info_parts.next()? != "3" {
+                return None;
+            }
+            let minor_version_string = version_info_parts.next()?;
+            PythonMinorVersion::from_version_string(minor_version_string)
+        });
+
+        let metadata = Self {
+            venv_path,
+            base_executable_home_path,
+            include_system_site_packages,
+            minor_version,
+        };
+
+        tracing::trace!("Resolved metadata for virtual environment: {metadata:?}");
+        Ok(metadata)
+    }
+
+    /// Return a list of `site-packages` directories that are available from this virtual environment
+    ///
+    /// See the documentation for `site_packages_dir_from_sys_prefix` for more details.
+    pub fn site_packages_dirs(
+        &self,
+        system: &dyn System,
+    ) -> Result<Vec<SystemPathBuf>, SitePackagesDiscoveryError> {
+        let VirtualEnvironment {
+            venv_path,
+            base_executable_home_path,
+            include_system_site_packages,
+            minor_version,
+        } = self;
+
+        let minor_version = minor_version.as_ref().copied();
```

Comment:
I would prefer a `as_u8` method. `as_ref` in combination with `copied` reads strange.

Reply:
I'm not sure it's possible for me to add a new method to `std:: option::Option`

Reply:
Oh, it's just about converting from `&Option<T>` to `Option<T>`. You can do

```
let minor_version = *minor_version;
```

I would then probably inline the expression